Go: [BREAKING] explicit destruction for futures and transactions

gm42 opened the pull request:
* Explicit destruction for futures and transactions.
* Introduction of context for transactions and futures.
* Simplification of Future interface and internal value caching

**This is a breaking change**

This change makes the Go binding stop relying on Go's GC finalizers and instead use explicit Close() calls to release resources, which is more Go-idiomatic; additionally, `context.Context` is added to function signatures and supported through transaction cancellation.

NOTE: once merged this will be a breaking change for users because memory leaks would start appear on their FoundationDB clients, unless `Close()` is called for futures, transactions and range iterators.

# Code-Reviewer Section

- [x] The PR has a description, explaining both the problem and the solution.
- [ ] The description mentions which forms of testing were done and the testing seems reasonable.
- [x] Every function/class/actor that was touched is reasonably well documented.

I have not yet updated test code as I'd like to discuss this change first.

## For Release-Branches

If this PR is made against a release-branch, please also check the following:

- [ ] This change/bugfix is a cherry-pick from the next younger branch (younger `release-branch` or `main` if this is the youngest branch)
- [ ] There is a good reason why this PR needs to go into a release branch and this reason is documented (either in the description above or in a linked GitHub issue)

gm42 committed:
Go: [BREAKING] explicit destruction for futures and transactions

Stop relying on Go's GC finalizers and instead use explicit Close() calls
to release resources, which is more Go-idiomatic.
The Get() method is removed from RangeIterator.

NOTE: this would be a breaking change for users because without any code breakage
they would now have huge memory leaks on their FoundationDB clients; a release management
solution must be devised for this problem.

gm42 commented on bindings/go/src/fdb/range.go:
```diff
@@ -249,40 +248,7 @@ func (ri *RangeIterator) fetchNextBatch() {
-		return
-	}
-
-	kv = ri.kvs[ri.index]
```
Note: on next advance call, this slice is thrown away. I'd like to address this in this PR, or a separate one, by having `Advance()` return the whole slice.

gm42 replied:
I ended up switching from `Advance()`/`GetSlice()` to `NextBatch()`/`Get()`

johscheuer reviewed (commented):
Thanks for the PR! Could we update the test cases to make use of the new close method (and directly document there how to use them)?

gm42 committed:
Go: rename Advance() as NextBatch(), remove RangeIterator.Get()

Other functions renamed:
* GetSliceWithError() -> Get()
* GetSliceOrPanic() -> MustGet()

gm42 commented on bindings/go/src/fdb/transaction.go:
```diff
@@ -88,7 +88,9 @@ func (opt TransactionOptions) setOpt(code int, param []byte) error {
 
-func (t *transaction) destroy() {
+// Close will destroy the underlying transaction.
+// It must be called after all the transaction-associated futures have been closed.
```
In fact destroying the transaction causes the transaction (and all its futures) to be cancelled, so this needs some re-thinking/re-writing as well.

gm42 replied:
I will make some attempts and submit a change; I am thinking to return a "closer" function that caller can defer.

gm42 replied:
I've submitted a version of this; will update all tests code.

gm42 committed:
Go: introduce context for Transact/ReadTransact

Use a goroutine to automatically cancel transaction when context has an error.
Correctly close transactions by returning a function to caller.

gm42 commented:
@johscheuer I have fixed the build errors in the directory layer; please trigger CI again

gm42 commented:
Thanks for triggering the CI @jzhou77.

I guess next step is to discuss the changes

johscheuer reviewed (commented):
Thanks again for the PR, it's in my queue for next week!

johscheuer commented on bindings/go/src/fdb/range.go:
```diff
@@ -170,119 +181,105 @@ func (rr RangeResult) GetSliceOrPanic() []KeyValue {
 // constructed with the (RangeResult).Iterator method.
 //
-// You must call Advance and get a true result prior to calling Get or MustGet.
+// You must call NextBatch and get a true result.
```
`NextBatch` doesn't return a `bool`, so the comment must be adjusted.

johscheuer commented on bindings/go/src/fdb/range.go:
```diff
@@ -170,119 +181,105 @@ func (rr RangeResult) GetSliceOrPanic() []KeyValue {
+		return nil, err
+	}
+	// consider reading done if limit was reached
+	ri.more = ri.more || ri.lastBatchLen == ri.options.Limit
```
```suggestion
	ri.more = ri.more && (ri.options.Limit == 0 || ri.lastBatchLen < ri.options.Limit)
```

I believe the current code has a bug. We want to read "more" only if we haven't exceeded the limit yet (assuming that the limit is not `0`). In the current code if `ri.more` would be `true` we would continue even if we hit the limit and run another `doGetRange` with the limit being set to `0` in this case (because of the `ri.prefetchOptions.Limit -= ri.lastBatchLen`).

gm42 replied:
I have rewritten it using an `if` branch, hopefully it helps with clarity?

johscheuer commented on bindings/go/src/fdb/range.go:
```diff
@@ -170,119 +181,105 @@ func (rr RangeResult) GetSliceOrPanic() []KeyValue {
+// operations associated with this result did not successfully complete.
+// If 'prefetch' is true, future for the next batch will be created on return and
+// released when the RangeIterator is closed.
+func (ri *RangeIterator) NextBatch(prefetch bool) ([]KeyValue, error) {
```
I'm fine with the renaming @jzhou77 any thoughts on the renaming?

gm42 replied:
I was looking at Java iterators for inspiration; if you have better ideas happy to change the implementation accordingly.

johscheuer commented on bindings/go/src/fdb/fdb_test.go:
```diff
@@ -127,37 +135,46 @@ func TestEstimatedRangeSize(t *testing.T) {
 	fdb.MustAPIVersion(API_VERSION)
 	db := fdb.MustOpenDefault()
-	_, err := db.ReadTransact(func(rtr fdb.ReadTransaction) (interface{}, error) {
+	_, txClose, err := db.ReadTransact(context.Background(), func(rtr fdb.ReadTransaction) (interface{}, error) {
```
Shouldn't we call in the method body `defer .Close()` on the returned future from the `rtr.Get` method? Otherwise we have a text example that introduces a memory leak?

johscheuer commented on bindings/go/src/fdb/fdb_test.go:
```diff
@@ -193,25 +210,33 @@ func ExampleReadTransactor() {
 	getOne := func(rt fdb.ReadTransactor, key fdb.Key) ([]byte, error) {
 		fmt.Printf("getOne called with: %T\n", rt)
-		ret, err := rt.ReadTransact(func(rtr fdb.ReadTransaction) (interface{}, error) {
+		ret, txClose, err := rt.ReadTransact(context.Background(), func(rtr fdb.ReadTransaction) (interface{}, error) {
```
Shouldn't we call in the method body `defer .Close()` on the returned future from the `rtr.Get` method? Otherwise we have a text example that introduces a memory leak?

johscheuer reviewed (commented):
I think it's good to add the explicit `close` method to the transaction(s). For the futures I'm not sure if it's a good idea or if we should at least add a helper method to get the value and close the future directly otherwise users will write repeatedly the same code where they get the future, get the value from it and then close the future.

gm42 commented:
> For the futures I'm not sure if it's a good idea or if we should at least add a helper method to get the value and close the future directly otherwise users will write repeatedly the same code where they get the future, get the value from it and then close the future.

I see what you mean; something like `GetAndClose()`? I will add it; the downside is that users might do something like this (I consider it an antipattern):
```go
		ret, txClose, err := rt.ReadTransact(context.Background(), func(rtr fdb.ReadTransaction) (interface{}, error) {
			values := make([]byte, 1000)
			for i := 0; i < 1000; i++ {
				key := fdb.Key(fmt.Sprintf("my-key-%d", i))
				values[i] = rtr.GetAndClose(key)
			}
			return values, nil
		})
		if err != nil {
			return nil, err
		}
		txClose()
```

instead of:
```go
		futures := make([]fdb.FutureByteSlice, 1000)
		ret, txClose, err := rt.ReadTransact(context.Background(), func(rtr fdb.ReadTransaction) (interface{}, error) {
			for i := 0; i < 1000; i++ {
				key := fdb.Key(fmt.Sprintf("my-key-%d", i))
				futures[i] = rtr.Get(key)
			}
			return nil, nil
		})
		if err != nil {
			return nil, err
		}

		// ... do something else while values are being fetched ...

		values := make([]byte, 1000)
		for i:=0;i<1000;i++ {
			values[i], err = futures[i].Get()
			futures[i].Close()
			if err != nil {
				return nil, err
			}
		}
		txClose()
```

Simply because the latter requires more code/typing.

Semisol reviewed (commented):
The future on line `137` of `allocator.go` is not closed. (`candidateValue`)

I also noticed that if there is a panic inside of a transaction, the transaction will not be cleaned up, as the automatic cancel only runs if there is an error *returned* and not a panic.

(GitHub won't let me comment on the specific line as it is not a part of the diff.)

gm42 commented:
> I also suggest that the Future APIs should be reworked so that `MustGet` or `Get` close the future.

Yes, I reached the same conclusion last week.

> I think the following would be best:

thanks for the feedback @Semisol!

> For parity, all futures' `Get` (except internal ones) are updated to cache return values.

This seems doable, I will try to add it to this PR.

> Each future has a new sync.RWMutex guarding f.ptr. f.ptr is set to nil after closure. This is only relevant for internal use

Not sure about this one; will try to find out whether it's the only option or not, as I change more code.

> The Destroy and Cancel methods are merged into Cancel.

I think you mean Close and Cancel; however, if you look at the C API there are specific behaviors associated to them. Will eventually remove `Destroy` if it's not necessary anymore for the user to call it.

> * `Get` will automatically close the future after values are ready.

Yes, this is doable thanks to the callback. I will instrument it further.

> * Developers are advised to do `Cancel` if they determine a future will not be used, or if they are not sure if it will (after any potential users go out of scope).

I think the current pattern with `Close()` matches what we do in Go in other cases (for example, with HTTP response bodies) and I would stick to it. Rather than asking developers to conditionally use `Cancel()`, it's safer to tell them to always use `Close()` and we have the method do nothing when there's nothing to do.

Will submit some changes later as I rewrite part of the current ones and add `context.Context` also to the future get functions

gm42 committed:
Go: add context to futures' Get()

Futures will be cancelled in case of context cancellation
Explicitly mention that BlockUntilReady is a reimplementation of fdb_future_block_until_ready

gm42 committed:
Go: use a channel instead of a Mutex to wait for future readiness

Using a channel allows an easier integration with context.Context

gm42 commented:
> Will submit some changes later as I rewrite part of the current ones and add `context.Context` also to the future get functions

I just submitted the first part of this rewrite:
* [x] add `context.Context` to future Getter functions
* [x] use a `sync.WaitGroup` instead of `sync.Mutex` to wait for callback (might eventually use a channel here when adding context support)

What's left to be done:
* [ ] actually use context in future Getter functions
* [ ] align futures to internally-cached approach
* [ ] release future when the callback re-enters
* [ ] always destroy instead of using `fdb_future_release_memory`

gm42 committed:
Go: cache resolved futures for all types

Only a few futures are currently being cached; this change
make all futures be cached and adds the missing check for fdb_future_get_error
in the previously cached futures.

gm42 committed:
Go: remove BlockUntilReady() and Cancel() from Future

Simplification for the usage of futures:
* if Get() is called on a future, all further calls will return a cached value and
  the future will be automatically closed
* if Get() is never called then future will be closed when user calls Close()

Cancel() is no more necessary: user can call Close()
Similarly, BlockUntilReady() is removed because a call to Get() (ignoring return value)
is equivalent.

gm42 commented:
I've completed my changes; a couple methods have been removed from the `Future` interface, I think this way it's easier to use it.

By the way, the C FDB API implementation does indeed treat destroy and cancel as almost the same:
```c++
extern "C" DLLEXPORT void fdb_future_cancel(FDBFuture* f) {
	CATCH_AND_DIE(TSAVB(f)->addref(); TSAVB(f)->cancel(););
}

extern "C" DLLEXPORT void fdb_future_destroy(FDBFuture* f) {
	CATCH_AND_DIE(TSAVB(f)->cancel(););
}
```

gm42 commented:
> I think the current design of that is quite iffy, abstracting a normal transaction, and a transaction _provider_ that does retries as one

I noticed that, but rather than an even bigger change I opted for adding context without pulling that apart. There is no performance/correctness concern other than it does not look elegant.

gm42 commented:
> I have done some of my own changes here, with some opiniation: https://github.com/semisol/foundationdb
> 
> I'll look into the changes here and get back with feedback soon.

I checked your branch; aside from undoing cancellation and context I think it introduces transaction leaks (with no finalizer and no closer, the transaction will not be closed). I saw you wanted to counteract that by closing the transaction in the `retryable`, but not sure that you can retry anything if the transaction has already been closed.

Looking forward to your feedback!

gm42 commented:
Let me write what is (at least for me) an interesting problem regarding closing transactions: there are basically two use cases to support:
1. opening a transaction, creating some futures, resolving them and getting their values before [commit](https://apple.github.io/foundationdb/api-c.html#c.fdb_transaction_commit) or [close](https://apple.github.io/foundationdb/api-c.html#c.fdb_transaction_destroy) (for RO transactions)
2. opening a transaction, creating some futures, committing the transaction, then resolving futures and getting their value, then closing the transaction.

In (2), closing the transaction causes all futures to be cancelled, thus it can only be done after user is done using the futures.
I think (2) is a valid use-case and we should not stop supporting it because it allows to commit a transaction and transfer the resolved futures' values from C memory to Go memory only as needed, after the transaction commit. This allows to complete a commit very quickly and process the gathered results afterwards.

gm42 commented:
> `retryable` returns only when it is not possible to retry anymore, so closure in that case is correct.

It is not correct IMO because closing a transaction implicitly cancels all associated futures, which can be used after transaction commit.



> This should be avoided in most cases, as you cannot retry once you are outside of the transaction, except with application logic. At that point, the user should use the `CreateTransaction` API and integrate it into their application logic.
> 
> I am not against supporting such an API though, but I believe it should be distinct.

Futures are perfectly valid to be used after transaction has been committed; this is currently supported when using the C binding and when using the Go binding. There is no need to retry because resolving a future doesn't involve any future-specific retry.

gm42 commented:
The failure seems to be:
```
clang: warning: argument unused during compilation: '-L/codebuild/output/src516627942/src/github.com/apple/foundationdb/build_output/fdbclient/awssdk-build/install/external-install/zlib/lib' [-Wunused-command-line-argument]
/codebuild/output/src516627942/src/github.com/apple/foundationdb/build_output/fdbclient/awssdk-build/curl/CMake/CurlTests.c:322:24: error: call to undeclared function 'ioctlsocket'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  322 |  unsigned long flags = ioctlsocket(socket, FIONBIO, &flags);
      |                        ^
/codebuild/output/src516627942/src/github.com/apple/foundationdb/build_output/fdbclient/awssdk-build/curl/CMake/CurlTests.c:322:44: error: use of undeclared identifier 'FIONBIO'
  322 |  unsigned long flags = ioctlsocket(socket, FIONBIO, &flags);
      |                                            ^
2 errors generated.
ninja: build stopped: subcommand failed.
```

Unrelated to this PR? I will rebase it just in case it helps.

gm42 commented:
@jzhou77 can you please help me figure out whether the CI failure is legit or a false positive?

Semisol reviewed (changes requested):
In directory_layer.go, opening a directory is intentionally not parallelizable. What you did would convert an open on a path `[]string{"a", "b", "c"}` to trying to open on root `a`, `b` and `c`, not nested opens.

Also, I am getting some panics in production with the context cancellation behavior, which I removed in my own fork. I think the `Transactor` interface is a mistake in general, as it implies some sort of "isolation" and cancellability, even though any failure *must* cancel the entire transaction.

gm42 commented:
I'd still like to finish this, if there are some reviews.

brownleej commented on bindings/go/src/fdb/database.go:
```diff
@@ -214,36 +240,37 @@ func retryable(wrapped func() (interface{}, error), onError func(Error) FutureNi
 // See the Transactor interface for an example of using Transact with
 // Transaction and Database objects.
-func (d Database) Transact(f func(Transaction) (interface{}, error)) (interface{}, error) {
+func (d Database) Transact(ctx context.Context, f func(Transaction) (interface{}, error)) (interface{}, func(), error) {
```
Since the Transact function is responsible for most of the transaction lifecycle, including the commit, would it be possible to have the Transact function take responsibility for closing the transaction as well?

brownleej commented on bindings/go/src/fdb/futures.go:
```diff
@@ -136,34 +127,37 @@ type FutureByteSlice interface {
 	// successfully complete. The current goroutine will be blocked until the
 	// future is ready.
-	Get() ([]byte, error)
+	Get(context.Context) ([]byte, error)
```
My suspicion is that it will be relatively common for people to want to immediately call `Get` on a future that is returned by another function, as in some of the examples that previous called `tr.Commit().Get()`. With this change, it seems like that will always require storing the future in a temporary variable. What are the trade-offs around having `Get` automatically close the future, or having a variant of `Get` that automatically closes it? Do we think the most common case will be for people to want to perform other functions on the future after `Get`, or for `Get` to be the last function called?

gxglass reviewed (commented):
AI review attached.  I generally agree with the concern that without a migration plan a breaking change like this seems like something we can't just unleash on unsuspecting users.

[pr-12052-review.md](https://github.com/user-attachments/files/30206413/pr-12052-review.md)
